packages/modules/vehicles/tronity/api.py

```diff
@@ -0,0 +1,98 @@
+#!/usr/bin/env python3
+import logging
+import jwt
+import paho.mqtt.publish as publish
+import json
+
+from modules.common import req
+from modules.vehicles.tronity.config import TronityVehicleSocConfiguration, TronityVehicleSoc
+from modules.common.abstract_soc import SocUpdateData
+from modules.common.component_state import CarState
+from datetime import datetime
+
+log = logging.getLogger(__name__)
+
+
+def fetch_soc(config: TronityVehicleSocConfiguration, soc_update_data: SocUpdateData, vehicle: int) -> CarState:
+    log.debug("Fetching Tronity SOC")
+    session = create_session(config, vehicle)
+    tronity_vehicle_id = str(config.vehicle_id)
+    tronity_data = fetch_soc_for_vehicle(tronity_vehicle_id, session)
+    return CarState(soc=tronity_data['level'], range=tronity_data['range'], soc_timestamp=tronity_data['lastUpdate'])
+
+
+def is_token_valid(access_token: str) -> bool:
+    if not access_token:
+        log.debug("No token found")
+        return False
+
+    decoded_data = jwt.decode(jwt=access_token, verify=False, algorithms=['HS256'], options={"verify_signature": False})
+    if datetime.utcfromtimestamp(decoded_data['exp']) < datetime.utcnow():
+        log.debug("Token expired: %s", decoded_data)
+        return False
+    else:
+        log.debug("Token still valid: %s", decoded_data)
+        return True
+
+
+def write_token_mqtt(topic: str, token: str, config: TronityVehicleSocConfiguration):
+    try:
+        config.access_token = token
+        value: TronityVehicleSoc = TronityVehicleSoc(configuration=config)
+        value_to_mqtt = json.dumps(value.__dict__, default=lambda o: o.__dict__)
+        publish.single(topic, value_to_mqtt, hostname="localhost", port=1883, retain=True)
```

Comment:
```suggestion
        Pub().pub(topic, value_to_mqtt)
```
Mit der Pub-Klasse aus den helpermodules werden die Index-Parameter zentral gesetzt und das Publishen an den Broker erfolgt von einer zentralen Stelle im Code.

Reply:
Ich hatte an anderer Stelle bereits ein ähnliches Problem und konnte das am Ende nur über publish.single lösen. Bin mir sicher dass es hier auch noch einen anderen Weg gibt.

Wenn ich pub().pub nutze wird das bereits JSON-Formatierte Object erneut als JSON formatiert und somit unbrauchbar. So landet dass dann im Broker:
`"\"{\"name\": \"Tronity\", \"type\": \"tronity\", \"configuration\": {\"vehicle_id\": \"64da870e4dc2130012f96f6b\"`

Lasse ich Zeile 42 weg kann ich den Wert nicht speichern da TronityVehicleSoc nicht JSON serialisierbar ist.
`2023-08-23 10:11:31,164 - {modules.vehicles.tronity.api:48} - {ERROR:fetch soc_ev3} - Token mqtt write exception Object of type TronityVehicleSoc is not JSON serializable
Traceback (most recent call last):
  File "/var/www/html/openWB/packages/modules/vehicles/tronity/api.py", line 46, in write_token_mqtt
    Pub().pub(topic, TronityVehicleSoc(configuration=config))
  File "/var/www/html/openWB/packages/helpermodules/pub.py", line 20, in pub
    self.publisher.client.publish(topic, payload=json.dumps(payload), qos=qos, retain=retain)
  File "/usr/lib/python3.9/json/__init__.py", line 231, in dumps
    return _default_encoder.encode(obj)
  File "/usr/lib/python3.9/json/encoder.py", line 199, in encode
    chunks = self.iterencode(o, _one_shot=True)
  File "/usr/lib/python3.9/json/encoder.py", line 257, in iterencode
    return _iterencode(o, 0)
  File "/usr/lib/python3.9/json/encoder.py", line 179, in default
    raise TypeError(f'Object of type {o.__class__.__name__} '
TypeError: Object of type TronityVehicleSoc is not JSON serializable`

Reply:
`from dataclass_utils import asdict
Pub().pub(topic, asdict(value_to_mqtt))`
Dafür ist das dataclass_utils bei den helpermodules gedacht. Die Klasse wird in ein Dict umgewandelt und wird an die pub-Methode übergeben und kann dann in json umgewandelt werden.